### PR TITLE
Rest API Support for zenbot

### PR DESCRIPTION
**Pull request by abduegal**
Hi guys,

Here is the code, that enables zenbot to have a REST API.
I tried to keep the code extendable, by allowing zenbot to add multiple **outputs** (one of them is this REST API). I have other **outputs** available like storing the trade in AWS dynamoDB. 
I also think it would be a good idea to refactor the stdout reporting part of zenbot, so that would also become an **output**.

Please let me know if you have any questions.

### CODE REVIEW

**Comment by storm1er** [comment]
Can you give us an exemple about how to use your output extension ? =)
_The same question goes for the REST API. (README.md appending ?)_

This PR feels like "+16843214 up" ... You could split this in some PR next time ^^'

File: extensions/output/api.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+module.exports = function container (get) {
+  let c = get('conf')
+  let express = require('express')
+  let app = express()
+  let random_port = require('random-port')
+
+  let run = function(reporter, tradeObject) {
+    if (!reporter.port || reporter.port === 0) {
+      random_port({from: 20000}, function(port) {
+        startServer(port, tradeObject)
```
Comment (DeviaVir) [inline extensions/output/api.js:12]: try/catch if the port doesn't work, and we try a different one automatically?

File: extensions/output/api.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      })
+    } else {
+      startServer(reporter.port, tradeObject)
+    }
+  }
+
+  let startServer = function(port, tradeObject) {
+    tradeObject.port = port
+
+    app.get('/tradedata', function (req, res) {
```
Comment (DeviaVir) [inline extensions/output/api.js]: `/trades`?

File: extensions/output/api.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      res.send(tradeObject)
+    })
+
+    app.listen(port)
+    tradeObject.url = require('ip').address() + ':' + port + '/tradeData'
```
Comment (DeviaVir) [inline extensions/output/api.js]: `/trades`?

**Review by DeviaVir — COMMENTED** [review]
I'm not sure if `lib/output.js` is the right place for this, but have no other suggestions at this time. I'll let others take a look and am interested in their feedback

File: extensions/output/api.js (hunk 2 above)
Comment (abduegal) [reply extensions/output/api.js]: agreed, I'll change that

File: extensions/output/api.js (hunk 1 above)
Comment (abduegal) [reply extensions/output/api.js:12]: okay I will do that

**Comment by abduegal** [comment]
@storm1er the output extension works the same as the notifiers, exchanges and strategies extensions. For that reason I did not add any manual for it.

With regards to the rest api. The config-sample.js should be self explanatory. I will add the available url and path names in a readme. 

Also, I do not know how I can split this pull request even further, all it contains is REST api support for zenbot.

File: extensions/output/api.js (hunk 1 above)
Comment (abduegal) [reply extensions/output/api.js:12]: the random_port function already searches for a free port, so I dont think this is neccesary anymore. The function also tests explicitly if the port is available through a try / catch mechanism

Comment (abduegal) [reply extensions/output/api.js:12]: image

Comment (DeviaVir) [reply extensions/output/api.js:12]: Ah, sweet! Thanks for digging that up.

**Review by DeviaVir — APPROVED** [review]
:+1: LGTM now. Thanks for this great addition!

**Comment by albabosh** [comment]
Seems that REST API requires two ports. If I put a single port in the config the app crashes a start. Could you give me a working config example ?

**Comment by storm1er** [comment]
@albabosh I don't have any issues with this PR,
Can you explain what's your problem ?

**Comment by albabosh** [comment]
If I enable REST API in conf.js & set c.output.api.port = 20002 it crashes a start (seems that express tries to bind twice to one port). I I set c.output.api.port = 0 zenbot starts well and binds to 2 random ports. I need to specify those two ports in confg (not random)

**Comment by storm1er** [comment]
hmm That's weird ... I didn't need have this problem ... Didn't need to manually set port anyway ^^'
Try to look what they said in review, probably a good start

**Pull request merged**